tensorflow_addons/layers/normalizations.py

```diff
@@ -316,3 +316,225 @@ def __init__(self, **kwargs):
+        self.gamma_regularizer = tf.keras.regularizers.get(gamma_regularizer)
+        self.tau_regularizer = tf.keras.regularizers.get(tau_regularizer)
+        self.beta_constraint = tf.keras.constraints.get(beta_constraint)
+        self.gamma_constraint = tf.keras.constraints.get(gamma_constraint)
+        self.tau_constraint = tf.keras.constraints.get(tau_constraint)
+        self.eps_var_initializer = tf.keras.initializers.get(
+            eps_var_initializer)
+        self.supports_masking = True
+        if isinstance(axis, list):
+            self.axis = axis[:]
+        else:
+            raise TypeError('axis must be list, type given: %s' % type(axis))
+        self._check_axis()
+
+    def build(self, input_shape):
+        self._check_if_input_shape_is_none(input_shape)
+        self._create_input_spec(input_shape)
+        self._add_gamma_weight(input_shape)
+        self._add_beta_weight(input_shape)
+        self._add_tau_weight(input_shape)
+        self._add_eps_var_weight()
+        self.built = True
+        super(FilterResponseNormalization, self).build(input_shape)
+
+    def compute_output_shape(self, input_shape):
+        return input_shape
+
+    def get_config(self):
+        config = {
+            'axis':
+            self.axis,
+            'epsilon':
+            self.epsilon,
+            'beta_initializer':
+            tf.keras.initializers.serialize(self.beta_initializer),
+            'gamma_initializer':
+            tf.keras.initializers.serialize(self.gamma_initializer),
+            'tau_initializer':
+            tf.keras.initializers.serialize(self.tau_initializer),
+            'eps_var_initializer':
+            tf.keras.initializers.serialize(self.eps_var_initializer),
+            'beta_regularizer':
+            tf.keras.regularizers.serialize(self.beta_regularizer),
+            'gamma_regularizer':
+            tf.keras.regularizers.serialize(self.gamma_regularizer),
+            'tau_regularizer':
+            tf.keras.regularizers.serialize(self.tau_regularizer),
+            'beta_constraint':
+            tf.keras.constraints.serialize(self.beta_constraint),
+            'gamma_constraint':
+            tf.keras.constraints.serialize(self.gamma_constraint),
+            'tau_constraint':
+            tf.keras.constraints.serialize(self.tau_constraint)
+        }
+        base_config = super(FilterResponseNormalization, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+    def call(self, inputs):
+        # Compute the mean norm of activations per channel.
+        nu2 = tf.reduce_mean(tf.square(inputs), axis=self.axis, keepdims=True)
+        # Perform FRN.
+        inputs = inputs * tf.math.rsqrt(nu2 + tf.math.abs(self.epsilon))
+        # Return after applying the Offset-ReLU non-linearity.
+        return tf.maximum(self.gamma * inputs + self.beta,
+                          self.tau + self.eps_var)
```

Comment:
In the paper, there's only a single use of `epsilon`. Which can either be a constant value of `1e-6` **or** can be a learnable param with the following form: `eps = 1e-6 + abs(eps_l)` where `eps_l` is the param with initial value of `1e-4`. 

You've used two different epsilons: `self.epsilon` (which is constant) and a `self.eps_var` which is a param with zero init. Secondly, the use of `eps` in the calculation of `FRN` is okay. But in the paper, there's no use of `eps` in the TLU whereas you've done:
```
z = max(y, tau + eps)
```
instead of 
```
z = max(y, tau)
```

Reply:
In most of the cases, the default value of `epsilon=1e-6` would work but not for the architectures where 1x1 conv is involved. In that case, the authors add another learnable parameter `eps_l`, which should be non-negative, so that the optimal epsilon value can be found out. 

Regarding `TLU`, yes that's a typo. Actually the  `eps_var` had to be added to the `epsilon` in the above line. I will correct it. Thanks for looking into it

Reply:
Another thing we can do is keep only `epsilon` as the trainable parameter with a constraint that it should only be positive and initialized with `1e-6`. Let me know what you think.

Reply:
No, that is not a good idea and it did not work well. For applications where it needs to be learnable, the parameterization of 1e-6 + abs(eps_l) worked well (where eps_l was initialized to 1e-4 or larger). Initializing it to small value tended to blow up the learning.

Reply:
So, we need the constant value of `epsilon=1e-6` as well as the learnable `eps_l` initialized to 1e-4.

Reply:
Right, the constant value lower bounds the value of effective epsilon, while the higher initialization prevent blow up at the start of the training.